core/src/block_creation_loop.rs

```diff
@@ -0,0 +1,521 @@
+        poh_recorder,
+        leader_schedule_cache,
+        rpc_subscriptions,
+        banking_tracer,
+        slot_status_notifier,
+        leader_window_notifier,
+        replay_highest_frozen,
+        record_receiver_receiver,
+    } = config;
+
+    // Similar to Votor, if this loop dies kill the validator
+    let _exit = Finalizer::new(exit.clone());
+
+    // get latest identity pubkey during startup
+    let mut my_pubkey = cluster_info.id();
+
+    info!("{my_pubkey}: Block creation loop initialized");
+
+    // Wait for PohService to be shutdown
+    let record_receiver = match record_receiver_receiver.recv() {
+        Ok(receiver) => receiver,
+        Err(e) => {
+            error!("{my_pubkey}: Failed to receive RecordReceiver from PohService. Exiting: {e:?}",);
+            return;
+        }
+    };
+
+    info!("{my_pubkey}: Block creation loop starting");
+
+    let mut ctx = LeaderContext {
+        exit,
+        my_pubkey,
+        leader_window_notifier: leader_window_notifier.clone(),
+        blockstore,
+        poh_recorder: poh_recorder.clone(),
+        record_receiver,
+        leader_schedule_cache,
+        bank_forks,
+        rpc_subscriptions,
+        slot_status_notifier,
+        banking_tracer,
+        replay_highest_frozen,
+    };
+
+    // Setup poh
+    // Important this is called *before* any new alpenglow
+    // leaders call `set_bank()`, otherwise, the old PoH
+    // tick producer will still tick in that alpenglow bank
+    // Since we received the `record_receiver` from PohService, we know that PohService has shutdown
+    // AND that replay no longer touches poh recorder. At this point BlockCreationLoop is the sole
+    // modifier of poh_recorder
+    {
+        let mut w_poh_recorder = ctx.poh_recorder.write().unwrap();
+        w_poh_recorder.enable_alpenglow();
+    }
+    reset_poh_recorder(&ctx.bank_forks.read().unwrap().working_bank(), &ctx);
+
+    while !ctx.exit.load(Ordering::Relaxed) {
+        // Check if set-identity was called at each leader window start
+        if my_pubkey != cluster_info.id() {
+            let my_old_pubkey = my_pubkey;
+            my_pubkey = cluster_info.id();
+            ctx.my_pubkey = my_pubkey;
+
+            warn!(
+                "Identity changed from {my_old_pubkey} to {my_pubkey} during block creation loop"
+            );
+        }
+
+        // Wait for the voting loop to notify us
+        let LeaderWindowInfo {
+            start_slot,
+            end_slot,
+            parent_block: (parent_slot, _),
+            skip_timer,
+        } = {
+            let window_info = leader_window_notifier.window_info.lock().unwrap();
+            let (mut guard, timeout_res) = leader_window_notifier
+                .window_notification
+                .wait_timeout_while(window_info, Duration::from_secs(1), |wi| wi.is_none())
+                .unwrap();
+            if timeout_res.timed_out() {
+                continue;
```

Comment:
is the timeout just so we can check for identity change? Might be nice to drop a comment here if so